lib/spack/spack/util/download.py

```diff
@@ -0,0 +1,388 @@
+    path: str = "",
+) -> DownloadInfo:
+    """Create a DownloadInfo object from a RequestInfo object."""
+    request = RequestInfo(
+        url=url, effective_url=effective_url, headers=headers or http.client.HTTPMessage()
+    )
+    return DownloadInfo(request=request, path=path)
+
+
+class UrlStreamReader:
+    """Context manager that reads from an URL stream."""
+
+    def __init__(self, *, url: str, timeout: int):
+        self.url = url
+        self.timeout = timeout
+
+    def __enter__(self) -> "UrlStreamReader":
+        raise NotImplementedError
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def read(self, size: int) -> bytes:
+        """Reads and returns a chunk of the response body."""
+        raise NotImplementedError
+
+    def request_info(self) -> RequestInfo:
+        """Returns information about the current request."""
+        raise NotImplementedError
+
+
+class UrllibStreamReader(UrlStreamReader):
+    """Streams a file from a URL using urllib."""
+
+    def __enter__(self) -> UrlStreamReader:
+        request = urllib.request.Request(self.url, headers={"User-Agent": SPACK_USER_AGENT})
+        self._response = urlopen(request, timeout=self.timeout)
+        return self
+
+    def read(self, size: int) -> bytes:
+        chunk = self._response.read(size)
+        return chunk
+
+    def request_info(self) -> RequestInfo:
+        effective_url = self.url
+        if isinstance(self._response, http.client.HTTPResponse):
+            effective_url = self._response.geturl()
+        return RequestInfo(
+            url=self.url, effective_url=effective_url, headers=self._response.headers
+        )
+
+
+class CurlStreamReader(UrlStreamReader):
+    """Streams a file from a URL using curl."""
+
+    _curl_exe: Optional[str] = None
+
+    def __init__(self, *, url: str, timeout: int, config_args: Optional[List[str]] = None):
+        super().__init__(url=url, timeout=timeout)
+        self._config_args: List[str] = config_args or []
+        if CurlStreamReader._curl_exe is None:
+            CurlStreamReader._curl_exe = which_string("curl", required=True)
+        self._curl = CurlStreamReader._curl_exe
+
+    def __enter__(self) -> UrlStreamReader:
+        curl_args = (
+            base_curl_fetch_args(
+                self.url, self.timeout, status_bar=False, user_agent=SPACK_USER_AGENT
+            )
+            + self._config_args
+        )
+        curl_cmd = [self._curl] + curl_args
+        self._curl_process = subprocess.Popen(
+            curl_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+
+        self._scheme = urllib.parse.urlparse(self.url).scheme
+        assert self._curl_process.stdout is not None, "curl process stderr is None"
+        self._stream = io.BufferedReader(self._curl_process.stdout)  # type: ignore[type-var]
+
+        # curl echoes intermediate redirect responses, so we might get multiple headers
```

Comment:
you always get multiple headers, you mean you get multiple responses.

Reply:
Done